src/jllvm/materialization/CodeGeneratorUtils.hpp

```diff
@@ -0,0 +1,197 @@
+                                                      llvm::StringRef methodType, ClassLoader& classLoader);
+
+public:
+    LazyClassLoaderHelper(ClassLoader& classLoader, llvm::orc::JITDylib& mainDylib, llvm::orc::JITDylib& implDylib,
+                          llvm::orc::IndirectStubsManager& stubsManager,
+                          llvm::orc::JITCompileCallbackManager& callbackManager, llvm::orc::IRLayer& baseLayer,
+                          llvm::orc::MangleAndInterner& interner, const llvm::DataLayout& dataLayout)
+        : m_mainDylib(mainDylib),
+          m_implDylib(implDylib),
+          m_stubsManager(stubsManager),
+          m_callbackManager(callbackManager),
+          m_baseLayer(baseLayer),
+          m_dataLayout(dataLayout),
+          m_classLoader(classLoader),
+          m_interner(interner)
+    {
+        m_mainDylib.withLinkOrderDo([&](const llvm::orc::JITDylibSearchOrder& dylibSearchOrder)
+                                    { m_implDylib.setLinkOrder(dylibSearchOrder); });
+    }
+
+    /// Creates a non-virtual call to the possibly static function 'methodName' of the type 'methodType' within
+    /// 'className' using 'args'. This is used to implement `invokestatic` and `invokespecial`.
+    llvm::Value* doNonVirtualCall(llvm::IRBuilder<>& builder, bool isStatic, llvm::StringRef className,
+                                  llvm::StringRef methodName, llvm::StringRef methodType,
+                                  llvm::ArrayRef<llvm::Value*> args);
+
+    enum MethodResolution
+    {
+        /// 5.4.3.3. Method Resolution from the JVM Spec.
+        Virtual,
+        /// 5.4.3.4. Interface Method Resolution from the JVM Spec.
+        Interface
+    };
+
+    /// Creates a virtual call to the function 'methodName' of the type 'methodType' within 'className' using 'args'.
+    /// 'resolution' determines how the actual method to be called is resolved using the previously mentioned strings.
+    llvm::Value* doIndirectCall(llvm::IRBuilder<>& builder, llvm::StringRef className, llvm::StringRef methodName,
+                                llvm::StringRef methodType, llvm::ArrayRef<llvm::Value*> args,
+                                MethodResolution resolution);
+
+    /// Returns an LLVM integer constant which contains the offset of the 'fieldName' with the type 'fieldType'
+    /// within the class 'className'.
+    llvm::Value* getInstanceFieldOffset(llvm::IRBuilder<>& builder, llvm::StringRef className,
+                                        llvm::StringRef fieldName, llvm::StringRef fieldType);
+
+    /// Returns an LLVM integer containing the iTable offset in the lower 8 bits and the id of the interface, whose
+    /// iTable should be indexed into from the 9th bit onwards for the class indicated by 'fieldDescriptor',
+    /// the method named 'methodName' with the type 'typeDescriptor'.
+    llvm::Value* getITableIdAndOffset(llvm::IRBuilder<>& builder, llvm::Twine fieldDescriptor,
```

Comment:
IIRC this method is actually dead code now that I forgot to remove. Feel free to do so.